File: Algorithms/C/sorting/shell_sort.c
#include <stdio.h>


void shell_sort(int arr[], int size, int shell)
{
    for (int gap = size / shell; gap > 0; gap /= shell)
    {
        for (int i = gap; i < size; i++)
        {
            int temp = arr[i];
            int j;
            for (j = i; j >= gap && arr[j - gap] > temp; j -= gap)
            {
                arr[j] = arr[j - gap];
            }
            arr[j] = temp;
        }
    }
}

int main()
{
    int arr[] = {12, 34, 54, 2, 3};
    int n = sizeof(arr) / sizeof(arr[0]);

    shell_sort(arr, n, 2);

    for (int i = 0; i < n; i++)
    {
        printf("%d ", arr[i]);
    }

    return 0;
}

File: Algorithms/Python/sorting/shell_sort.py
from __future__ import annotations

from typing import TypeVar

T = TypeVar("T", int, float, str)


def shell_sort(array: list[T], shell: int) -> None:
    gap = len(array) // shell
    while gap > 0:
        for i in range(gap, len(array)):
            temp = array[i]
            j = i
            while j >= gap and array[j - gap] > temp:
                array[j] = array[j - gap]
                j -= gap
            array[j] = temp
        gap //= shell


if __name__ == "__main__":
    array = list("SORTEXAMPLE")
    shell_sort(array, 7)
    print(*array)  # Output: [2, 3, 12, 34, 54]
